File: modules/bin_reporting/src/clients.js
/**
 * Outbound adapters to other modules.
 *
 * Rules, identical in spirit to every other module's client layer:
 *   1. Target from an env var — the provider is swappable at deploy time.
 *   2. Hard timeout on every call.
 *   3. Never throw; return { ok, ... } and let the caller degrade.
 *
 * The worker_dashboard adapter below carries the adaptation cost for that
 * acquired module: Bearer auth, snake_case, and its domain-neutral vocabulary
 * (our bin id becomes its `job_ref`). Keeping that translation here is what
 * lets the purchased module stay unmodified and therefore resaleable.
 */

const WASTE_RECOGNITION_URL = process.env.WASTE_RECOGNITION_URL || "";
const WORKER_DASHBOARD_URL = process.env.WORKER_DASHBOARD_URL || "";
const CREW_AUTH_TOKEN = process.env.CREW_AUTH_TOKEN || "dev-fieldops-token";
const ANALYTICS_URL = process.env.ANALYTICS_URL || "";

const TIMEOUT_MS = Number(process.env.DEPENDENCY_TIMEOUT_MS || 2500);

async function request(url, options = {}) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
  try {
    const res = await fetch(url, { ...options, signal: controller.signal });
    const body = await res.json().catch(() => null);
    if (!res.ok) {
      return { ok: false, reason: `upstream_${res.status}`, body };
    }
    return { ok: true, data: body };
  } catch (err) {
    return { ok: false, reason: err.name === "AbortError" ? "timeout" : "unreachable" };
  } finally {
    clearTimeout(timer);
  }
}

/**
 * Classify a photo. Degradation: the bin is stored `unclassified` and can be
 * backfilled later via POST /api/v1/reports/:id/reclassify.
 */
export async function classify(imageBuffer, reference) {
  if (!WASTE_RECOGNITION_URL) return { ok: false, reason: "not_configured" };
  return request(`${WASTE_RECOGNITION_URL}/api/v1/classify`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      image_base64: imageBuffer.toString("base64"),
      reference,
    }),
  });
}

/**
 * Request dispatch from the acquired FieldOps module. Note the vocabulary
 * translation: our bin id is its job_ref.
 */
export async function requestAssignment({ binId, location, wasteType }) {
  if (!WORKER_DASHBOARD_URL) return { ok: false, reason: "not_configured" };
  return request(`${WORKER_DASHBOARD_URL}/v1/assignments`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${CREW_AUTH_TOKEN}`, // FieldOps' scheme, not ours
    },
    body: JSON.stringify({
      job_ref: binId,
      location,
      metadata: { waste_type: wasteType || null },
    }),
  });
}

/** Emit an analytics event. Degradation: metrics lose a data point. */
export async function emit(type, subjectId, payload = {}) {
  if (!ANALYTICS_URL) return { ok: false, reason: "not_configured" };
  return request(`${ANALYTICS_URL}/api/v1/events`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ type, subject_id: subjectId, source: "bin_reporting", payload }),
  });
}

export const dependencyConfig = () => ({
  waste_recognition: WASTE_RECOGNITION_URL || null,
  worker_dashboard: WORKER_DASHBOARD_URL || null,
  analytics_dashboard: ANALYTICS_URL || null,
  timeout_ms: TIMEOUT_MS,
});

File: hacquire/modules/bin_reporting/app/clients.py
"""Outbound adapters.

Same three rules as every client layer in the registry: targets from env vars,
hard timeout, never raise — return {"ok": bool, ...} and let the caller degrade.

The worker_dashboard adapter carries the adaptation for that ACQUIRED module:
Bearer auth, snake_case, and its domain-neutral vocabulary (our bin id becomes
its `job_ref`). Keeping that translation here is what lets the purchased
module stay unmodified and therefore resaleable.
"""
import base64
import os

import httpx

WASTE_RECOGNITION_URL = os.getenv("WASTE_RECOGNITION_URL", "")
WORKER_DASHBOARD_URL = os.getenv("WORKER_DASHBOARD_URL", "")
CREW_AUTH_TOKEN = os.getenv("CREW_AUTH_TOKEN", "dev-fieldops-token")
ANALYTICS_URL = os.getenv("ANALYTICS_URL", "")
TIMEOUT_S = float(os.getenv("DEPENDENCY_TIMEOUT_MS", "2500")) / 1000


async def _request(method: str, url: str, **kw) -> dict:
    try:
        async with httpx.AsyncClient(timeout=TIMEOUT_S) as client:
            r = await client.request(method, url, **kw)
            if r.status_code >= 400:
                return {"ok": False, "reason": f"upstream_{r.status_code}"}
            return {"ok": True, "data": r.json()}
    except httpx.TimeoutException:
        return {"ok": False, "reason": "timeout"}
    except Exception:
        return {"ok": False, "reason": "unreachable"}


async def classify(image: bytes, reference: str) -> dict:
    """Degradation: the bin is stored unclassified and can be backfilled later
    via POST /api/v1/reports/{id}/reclassify."""
    if not WASTE_RECOGNITION_URL:
        return {"ok": False, "reason": "not_configured"}
    return await _request("POST", f"{WASTE_RECOGNITION_URL}/api/v1/classify",
                          json={"image_base64": base64.b64encode(image).decode(),
                                "reference": reference})


async def request_assignment(bin_id: str, location: dict, waste_type: str | None) -> dict:
    """Degradation: the report stays 'reported' and can be dispatched later.

    Note the vocabulary translation — our bin id is its job_ref.
    """
    if not WORKER_DASHBOARD_URL:
        return {"ok": False, "reason": "not_configured"}
    return await _request(
        "POST", f"{WORKER_DASHBOARD_URL}/v1/assignments",
        headers={"Authorization": f"Bearer {CREW_AUTH_TOKEN}"},  # FieldOps' scheme
        json={"job_ref": bin_id, "location": location,
              "metadata": {"waste_type": waste_type}},
    )


async def emit(event_type: str, subject_id: str, payload: dict) -> dict:
    """Degradation: the report succeeds; metrics lose a data point."""
    if not ANALYTICS_URL:
        return {"ok": False, "reason": "not_configured"}
    return await _request("POST", f"{ANALYTICS_URL}/api/v1/events",
                          json={"type": event_type, "subject_id": subject_id,
                                "source": "bin_reporting", "payload": payload})


def dependency_config() -> dict:
    return {
        "waste_recognition": WASTE_RECOGNITION_URL or None,
        "worker_dashboard": WORKER_DASHBOARD_URL or None,
        "analytics_dashboard": ANALYTICS_URL or None,
        "timeout_ms": int(TIMEOUT_S * 1000),
    }
